lib/rack/request.rb

```diff
@@ -16,7 +16,20 @@ class << self
       attr_accessor :ip_filter
     end
 
-    self.ip_filter = lambda { |ip| /\A127\.0\.0\.1\Z|\A(10|172\.(1[6-9]|2[0-9]|30|31)|192\.168)\.|\A::1\Z|\Afd[0-9a-f]{2}:.+|\Alocalhost\Z|\Aunix\Z|\Aunix:/i.match?(ip) }
+    VALID_IPV4_OCTET = '(25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])'.freeze
+    
+    TRUSTED_PROXIES = Regexp.union(
+      /\A127(\.(#{VALID_IPV4_OCTET})){3}$\Z/,                         # localhost IPv4 range 127.x.x.x, per RFC-3330
```

Comment:
Is there a reason for `$\Z` instead of just `\Z`? If not, these can be changed to just `\Z`.

Reply:
Without the $ it is allowing the regexp to match when there is a trailing character, for example `127.0.0.300` matches without the $ because valid IP address `127.0.0.30` is inside of it.

Reply:
This sounded weird to me, so I tested it:

```ruby
VALID_IPV4_OCTET = '(25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])'
re = /\A127(\.(#{VALID_IPV4_OCTET})){3}$\Z/
re2 = /\A127(\.(#{VALID_IPV4_OCTET})){3}\Z/
re.match('127.0.0.300') # => nil
re2.match('127.0.0.300') # => nil 
```

The Ruby documentation for `\Z` states: `\Z - Matches end of string. If string ends with a newline, it matches just before newline`.  `0` is not a newline, so `\Z` should not match before the final `0`.

Note that in general, you should use `\z` (end of string), and not `\Z`.  `\Z` should only be used if you want to allow a trailing newline, but have the trailing newline not be matched.  So if you want to match against end of string, you should switch everything to `\z`.  My initial assumption from reading the code is you wanted to handle trailing newlines specially, since both `$` and `\Z` are related to newlines.  But from this discussion, it seems that was not the reason, and you should probably switch to `\z`.

Reply:
Converted all instances of `\Z` to `\z` and all tests still pass.

Reply:
Just `\z`, not `$\z`.  `$` before `\z` has no effect.

Reply:
Ah, I see, thank you.  I was using an external regexp tester to develop the patterns, and there must be a difference here between ruby and some other flavor like PCRE.

Removed the `$` and all tests still pass using only `\z`